[indirect-sender] new design for indirect tx and handling of data polls

abtink opened the pull request:
This PR changes the design of indirect frame tx in OpenThread core. The primary objective is move the handling of data polls and indirect tx closer to MAC layer. 
- This enables future enhancements to have RCP handle data polls and indirect tx (including retx of indirect frame tx).
- Also helps get the MAC layer definition closer (but not a 100% match) to 802.15.4 MCPS/MLME  style primitive definitions. 
- Helps simplify the code and reduce complexity and responsibility  of `MeshForwarder` module.
- Allows each module to define its own state/info it requires to be saved per child/neighbor in the `Child` class.

The PR contains multiple related commits each containing relatively smaller changes in the hope to structure the overall design and make it easier to review/read the code. Each commit is run through travis to ensure it builds correctly and that it passes all the tests on travis.

**[indirect-sender] adding new class IndirectSender**   
This commit adds a new module/class `IndirectSender` for handling
of the indirect transmission to sleepy children. The commit just
moves the existing code form `MeshForwarder` into the new class.

**[indirect-sender] define ChildInfo for child indirect tx related info**
This commit defines `IndirectSender::ChildInfo` class in child sender
module which contains all the info needed to store (per child) for
handling of indirect transmissions to the child. The `Child` class
inherits from the new class. This helps keep the indirect tx info in
the same file.

**[mac] add new operation for indirect data frame transmission**
This commit adds a new operation for indirect data frame transmission
in `Mac` class (only available on FTD build). The indirect tx request
is handled before a direct tx request. This commit also changes how
the number of retries and CMSA attempts (for a frame) are determined
`Mac` class itself now sets these according to request type.

**[mesh-forwarder] add PrepareDataFrame()**
This commit refactors and updates the existing code to add a generic
method `MeshForwarder::PrepareDataFrame()`. This method constructs a
MAC data from from a given IPv6 message at the offset set in the
message. This method enables link security when message is MLE type
and requires fragmentation.

**[data-poll-handler] adding DataPollHandler class**
This commit adds a new class `DataPollHandler` which sits between
`Mac` layer and `IndirectSender`. It interfaces to `Mac` to handle any
received data poll and perform indirect frame transmission including
the re-transmission logic of frames (per poll). All state info (per
child) for handling of indirect frame retransmission is now defined
and managed by the `DataPollHanlder` class itself. This commit updates
the `IndiretSender` to interface with the `DataPollHandler` class and
handle preparation of frames from `Message` for indirect transmission.

abtink commented:
Re-based on master and fixed the conflicts.

abtink commented on src/core/thread/mesh_forwarder.cpp:
```diff
@@ -827,16 +770,20 @@ otError MeshForwarder::SendFragment(Message &aMessage, Mac::Frame &aFrame)
                 aMessage.SetOffset(0);
-                ExitNow(error = OT_ERROR_NOT_CAPABLE);
+                aMessage.SetLinkSecurityEnabled(true);
+                goto start;
```
We use the same pattern of `goto` to a start point in [`Lowpan::Compress()`](https://github.com/openthread/openthread/blob/master/src/core/thread/lowpan.cpp#L250) from `lowpan.cpp`.

We can replace them with other options (recursive call, or wrap the entire block in a loop and use `continue` to jump to start of the loop), but I think the `goto` is more clear and readable (and shows the intention of the programmer).

I would favor that we relax the `goto` requirement in the style guide and change it to "avoid" from "should not" and selectively allow it. 

Thoughts? @jwhui?

abtink replied:
Submitted PR https://github.com/openthread/openthread/pull/3983 to update the style guide.

jwhui replied:
As mentioned in #3983, the avoidance of `goto` is currently a "should" that does allow limited exceptions.

abtink commented:
Thanks @CiaranWoodward.

Indeed, the goal is to make it easier to replace `DataPollHandler` while keeping `IndirectSender` same for different models (RCP or MCPS).

You are right on point on the limitation/assumption on how `mMessageNextOffset` is handled by `IndirectSender`. We were aware of this and considered some options. 

The current implementation of `IndirectSender` assumes that it would only get a `HandleSentFrameToChild()` callback after a `PrepareFrameForChild()` so it saves the next offset in a single member variable (this is similar to current behavior of `MeshForwarder`). This works perfectly with how `DataPollHandler`/`Mac` work currently but we definitely need to relax this for RCP handling data polls and/or MCPS model.

Ideally, we need to keep track of both a current and a next message offset for the the indirect message per child.

- Easiest solution is  to add a field for "next message offset" in the `ChildInfo`. Downside is that it would increase child entry size and child table overall. And  in the `DataPollHandler` case we don't practically need it since we will have only one active indirect tx at the time. 
- Another idea is to re-generate the frame in `HandleSentFrameToChild()` to derive the next offset again. Downside is increase processing time - trade-off memory/processing time -  I was not in favor of this.
- The solution I prefer is to delegate the saving of next offset to the lower layer module interfacing to `IndirectSender` (`DataPollHanlder` in this case or any future module interfacing to `IndirectSender`)
   - We let `IndirectSender` define a `FrameContxt` struct where it specifies any info it wants to save along with a prepared frame.
   - The lower-layer provides buffer for the context in `PrepareFrameForChild()` (same way it provides the frame buffer itself)
   - In `HandleSentFrameToChild()` the lower layer module includes the context info which is then used by the `IndirectSender`.
- I like this since it's flexible and efficient (we pay for what we use when we use it). 

I was originally thinking of including the context approach in this PR but decided separate it out and add it as a future PR.

Would be happy to hear your thoughts on this.

CiaranWoodward commented:
@abtink 
I agree with your proposed solution, passing a 'context' to the lower level to track the individual frame offsets is the same technique I settled on in my implementation. As you mention, this could then be 'paid for' in a manner that is suitable for the lower level.

abtink commented:
Rebased on master, and added `(#3952)` to commit msg title of all commits except the last one (@jwhui  I think GH will add it to last one when we merge but not earlier ones, is this correct?)

jwhui commented:
@abtink, thanks for the rebase. The PR reference should be applied to every commit we want to retain on a “rebase and merge”, including the last one.

jwhui commented on Android.mk:
```diff
@@ -153,6 +153,7 @@ LOCAL_SRC_FILES                                          := \
     src/core/crypto/sha256.cpp                              \
     src/core/mac/channel_mask.cpp                           \
     src/core/mac/data_poll_sender.cpp                       \
+    src/core/mac/data_poll_handler.cpp                      \
```
nit: move above `data_poll_sender.cpp`?

jwhui commented on src/core/Makefile.am:
```diff
@@ -157,6 +157,7 @@ SOURCES_COMMON                      = \
     crypto/sha256.cpp                 \
     mac/channel_mask.cpp              \
     mac/data_poll_sender.cpp          \
+    mac/data_poll_handler.cpp         \
```
nit: move above `data_poll_sender.cpp`?

jwhui commented on src/core/Makefile.am:
```diff
@@ -322,6 +324,7 @@ HEADERS_COMMON                      = \
     crypto/sha256.hpp                 \
     mac/channel_mask.hpp              \
     mac/data_poll_sender.hpp          \
+    mac/data_poll_handler.hpp         \
```
nit: move above `data_poll_sender.cpp`?

jwhui commented on src/core/thread/indirect_sender.cpp:
```diff
@@ -0,0 +1,508 @@
+
+void IndirectSender::Start(void)
+{
+    if (!mEnabled)
```
nit: is this check necessary?

jwhui commented on src/core/thread/mesh_forwarder.cpp:
```diff
@@ -610,38 +600,44 @@ otError MeshForwarder::HandleFrameRequest(Mac::Frame &aFrame)
 }
 
-otError MeshForwarder::SendFragment(Message &aMessage, Mac::Frame &aFrame)
+void MeshForwarder::SendFragment(Message &aMessage, Mac::Frame &aFrame)
```
Given that there is only one call to this, should we just call `PrepareDataFrame()` directly in `HandleFrameRequest()`?

abtink replied:
Good point. Will change it.

jwhui commented on src/core/mac/data_poll_handler.hpp:
```diff
@@ -0,0 +1,231 @@
+     *
+     * @retval OT_ERROR_NONE  The request was processed successfully.
+     * @retval OT_ERROR_BUSY  The request cannot be handled now. Upon handling of the request, the
+                              `Callbacks::UpdateIndirectMessage()` will be invoked to notify the caller.
```
nit: missing `*`

jwhui commented on src/core/mac/data_poll_handler.hpp:
```diff
@@ -0,0 +1,231 @@
+    /**
+     * This method processes a frame change request (new frame, replace frame, or purge frame) for a given child.
+     *
+     * When there is an ongoing indirect frame transmission to same child the request may not be handled immediately.
```
Rather than having the caller deal with calling `UpdateIndirectMessage()` on success, should this method simply handle that directly?

Also, the current method names do not clearly indicate the request/callback relationship (e.g. `ProcessFrameChange()` -> `UpdateIndirectMessage()`. What about having the method names share some common string that makes their relationship more obvious?

abtink replied:
Great idea. I updated it in the last set of pushed commits.

jwhui commented on src/core/mac/data_poll_handler.hpp:
```diff
@@ -0,0 +1,271 @@
+     */
+    enum FrameChange
+    {
+        kPurgeFrame,   // Indicates that previous frame should be purged and any ongoing indirect tx should be aborted.
```
nit: prefer `///<` for doxygen

jwhui commented on src/core/mac/data_poll_handler.hpp:
```diff
@@ -0,0 +1,271 @@
+         * now processed and the indirect message for the child can be updated.
+         *
+         * @param[in]  aChild     The child to update.
+         */
```
nit: add additional line

jwhui commented on src/core/mac/data_poll_handler.hpp:
```diff
@@ -0,0 +1,271 @@
+    explicit DataPollHandler(Instance &aInstance);
+
+    /**
+     * This method disables data poll handler and clears any state/info saved per child for indirect frame tx.
```
What does "disables data poll handler" mean?

abtink replied:
I changed the comment to mention it would only clear data (which is what it does).  Thanks

jwhui commented on src/core/thread/indirect_sender.hpp:
```diff
@@ -0,0 +1,223 @@
+        /**
+         * This method returns the number of queued message(s) for the child.
+         *
+         * @returns Number of queued message(s).
```
nit: "message(s)" -> "messages"
